server/src/main/scala/org/http4s/server/middleware/CSRF.scala

```diff
@@ -161,28 +207,101 @@ final class CSRF[F[_], G[_]] private[middleware] (
 
 object CSRF {
 
+  //Newtype hax
+  type CSRFToken <: String
+  private[CSRF] def lift(s: String): CSRFToken = s.asInstanceOf[CSRFToken]
+
   /** Default method for constructing CSRF middleware **/
   def apply[F[_]: Sync, G[_]: Applicative](
       headerName: String = "X-Csrf-Token",
       cookieName: String = "csrf-token",
       key: SecretKey,
-      clock: Clock = Clock.systemUTC()): CSRF[F, G] =
-    new CSRF[F, G](headerName, cookieName, key, clock)
+      clock: Clock = Clock.systemUTC(),
+      onFailure: Response[G] = Response[G](Status.Forbidden),
+      secure: Boolean = false,
+      createIfNotFound: Boolean = true,
+      headerCheck: Request[G] => Boolean): CSRF[F, G] =
+    new CSRF[F, G](
+      headerName,
+      cookieName,
+      key,
+      clock,
+      headerCheck,
+      onFailure,
+      secure,
+      createIfNotFound)
+
+  /** Default method for constructing CSRF middleware **/
+  def default[F[_]: Sync, G[_]: Applicative](
+      headerName: String = "X-Csrf-Token",
+      cookieName: String = "csrf-token",
+      key: SecretKey,
+      clock: Clock = Clock.systemUTC(),
+      onFailure: Response[G] = Response[G](Status.Forbidden),
+      host: String,
+      sc: Scheme,
+      port: Option[Int],
+      secure: Boolean = false,
+      createIfNotFound: Boolean = true): CSRF[F, G] =
+    new CSRF[F, G](
+      headerName,
+      cookieName,
+      key,
+      clock,
+      defaultOriginCheck(_, host, sc, port),
+      onFailure,
+      secure,
+      createIfNotFound)
+
+  /** Check origin matches our proposed origin.
+    *
+    * @param r
+    * @param host
+    * @param sc
+    * @param port
+    * @tparam F
+    * @return
+    */
+  def defaultOriginCheck[F[_]](
+      r: Request[F],
+      host: String,
+      sc: Scheme,
+      port: Option[Int]): Boolean =
+    r.headers
+      .get(Origin)
+      .flatMap(o => Uri.fromString(o.value).toOption)
+      .exists(u => u.host.exists(_.value == host) && u.scheme.contains(sc) && u.port == port) || r.headers
+      .get(Referer)
+      .exists(u =>
+        u.uri.host.exists(_.value == host) && u.uri.scheme.contains(sc) && u.uri.port == port)
+
+  def proxyOriginCheck[F[_]](r: Request[F], host: Host, xff: `X-Forwarded-For`): Boolean =
```

Comment:
Probably could use a convenience constructor?

Reply:
I mean a way to create a `CSRF` without forcing the person to use `apply` and pass this around